### PR TITLE
Add trim/0, ltrim/0 and rtrim/0 that trims leading and trailing whitespace

### PR DESCRIPTION
**Pull request by wader**

### CODE REVIEW

File: src/jv_unicode.c [hunk 1]
```diff
@@ -118,3 +118,16 @@ int jvp_utf8_encode(int codepoint, char* out) {
   assert(out - start == jvp_utf8_encode_length(codepoint));
   return out - start;
 }
+
+// space codepoints for unicode basic latin and latin-1 supplement
```
Comment (wader) [inline src/jv_unicode.c]: This is same as what golang `strings.TrimSpace`/`unicode.IsSpace` considers whitespace https://pkg.go.dev/unicode#IsSpace

File: docs/content/manual/manual.yml [hunk 2]
```diff
@@ -1772,6 +1772,27 @@ sections:
...
+
+          Whitespace is
+          `" "`,
+          `"\n"`
+          `"\t"`,
+          `"\r"`,
+          `"\f"`,
+          `"\u000b"` (vertical tab),
+          `"\u0085"` (next line) and
+          `"\u00a0"` (no-break space). These are the whitespace characters in
```
Comment (nicowilliams) [inline docs/content/manual/manual.yml]: Is this exhaustive for Unicode, or just Latin scripts in Unicode?  If the latter, why not be more exhaustive?

File: src/builtin.c [hunk 3]
```diff
@@ -1184,6 +1184,47 @@ static jv f_string_indexes(jq_state *jq, jv a, jv b) {
   return jv_string_indexes(a, b);
 }
 
+static jv f_string_trim(jq_state *jq, jv a) {
```
Comment (nicowilliams) [inline src/builtin.c]: We could jq-code `trim`...

File: docs/content/manual/manual.yml (hunk 2 above)
Comment (wader) [reply docs/content/manual/manual.yml]: Is only whitespace from basic block and latin-1, so not exhaustive. Reason is mostly to match what other implementations do. There are quite a lot of other whitespace characters in other blocks, wikipedia has a good list https://en.wikipedia.org/wiki/Whitespace_character.

Reading about PCRE's `\s`-class it seem to match something similar.

File: src/builtin.c (hunk 3 above)
Comment (wader) [reply src/builtin.c]: Main reason was i couldn't come up with something that i thought would be reasonably performant. Any ideas?   reduce/slice somehow? fq's trim is `def trim: capture("^\\s*(?<str>.*?)\\s*$"; "m").str;` atm but i fear that compiling a regex each time might not be great (maybe a regex cache would be good?)

Maybe time to measure?

File: docs/content/manual/manual.yml (hunk 2 above)
Comment (wader) [reply docs/content/manual/manual.yml]: Rust `trim`/`is_whitespace` uses characters with `White_Space` property https://www.unicode.org/Public/UCD/latest/ucd/PropList.txt maybe more resonable?

File: src/jv_unicode.c (hunk 1 above)
Comment (itchyny) [reply src/jv_unicode.c]: Note that Go's `strings.TrimSpace` considers `White_Space` property as well.

File: src/builtin.c (hunk 3 above)
Comment (nicowilliams) [reply src/builtin.c]: Good point.

File: docs/content/manual/manual.yml (hunk 2 above)
Comment (nicowilliams) [reply docs/content/manual/manual.yml]: Maybe we should document that the list is not stable, that we may add to it.

Comment (nicowilliams) [reply docs/content/manual/manual.yml]: > Rust `trim`/`is_whitespace` uses characters with `White_Space` property https://www.unicode.org/Public/UCD/latest/ucd/PropList.txt maybe more resonable?

I do like that better, yeah, especially in light of @itchyny's comment.

Comment (itchyny) [reply docs/content/manual/manual.yml]: Wait for utf8proc to be included for upcase/downcase?

Comment (wader) [reply docs/content/manual/manual.yml]: Update to `White_Space` but i guess we can wait. Also update docs and tests.

**Comment by itchyny** [comment]
I found [a list of languages on this function](https://en.wikipedia.org/wiki/Comparison_of_programming_languages_(string_functions)#trim). I have no objection to name this trim not strip, because we already have ltrimstr. Does anyone want ltrim and rtrim as well?

**Comment by wader** [comment]
> I found [a list of languages on this function](https://en.wikipedia.org/wiki/Comparison_of_programming_languages_(string_functions)#trim). I have no objection to name this trim not strip, because we already have ltrimstr. Does anyone want ltrim and rtrim as well?

Personally mostly have needed `trim`, and i agree that also having `ltrim`/`rtrim` would fits well with `ltrimstr`/`rtrimstr`.

**Comment by wader** [comment]
Added `ltrim/0` and `rtrim/0`

File: src/builtin.c [hunk 4]
```diff
@@ -1197,6 +1197,58 @@ static jv f_string_indexes(jq_state *jq, jv a, jv b) {
...
+
+static jv string_trim(jv a, int op) {
+  if (jv_get_kind(a) != JV_KIND_STRING) {
+    return ret_error(a, jv_string("trim input must be a string"));
+  }
+
+  int len = jv_string_length_bytes(jv_copy(a));
+  const char* start = jv_string_value(a);
+  const char* trim_start = start;
+  const char* end = trim_start + len;
```
Comment (emanuele6) [inline src/builtin.c]: `char *`

File: src/builtin.c [hunk 5]
```diff
@@ -1197,6 +1197,58 @@ static jv f_string_indexes(jq_state *jq, jv a, jv b) {
...
+  const char *trim_end = end;
+  int c;
+
+  if (op & TRIM_LEFT) {
+    for (;;) {
+      const char* ns = jvp_utf8_next(trim_start, end, &c);
```
Comment (emanuele6) [inline src/builtin.c]: `char *`

**Review by emanuele6 — APPROVED** [review]
Looks good to me

**Comment by wader** [comment]
@emanuele6 thanks for review! maybe wait for one more apporval? thinking adding new functions might be good idea with more consensus

About waiting for utf8proc from #2547: i think we can merge this separately and fix that later

**Review by itchyny — APPROVED** [review]
LGTM
After utf8proc is included later, let's see whether we can clean up (or get rid of) jv_unicode.c.

**Comment by wader** [comment]
> LGTM After utf8proc is included later, let's see whether we can clean up (or get rid of) jv_unicode.c.

Yeap, had a quick look at it a few week ago and it seemed like it would be easy, something like `utf8proc_get_property(c)->category == UTF8PROC_CATEGORY_ZS `

**Pull request merged**